### DISCUSSION TITLE
Changed Vitest 1.0.0 CLI breaks debugging in IntelliJ / WebStorm

### DISCUSSION THREAD

**Discussion by segrey** [vitest-dev/vitest#4643] [General]
CLI has changed in Vitest 1.0.0-beta resulting in broken Vitest debugging in IntelliJ / WebStorm ([WEB-64001](https://youtrack.jetbrains.com/issue/WEB-64001/Prepare-update-to-Vitest-1.0.0)), because IntelliJ / WebStorm still passes `--threads=false` when debugging tests as it did for Vitest 0.x.
Honestly, I don't remember why it passes `--threads=false` in the first place, because debugging seems to be working without this option.

What would the correct way to start Vitests for debugging? For example, all of the following ways work for me:
* no CLI options
* `--poolOptions.threads.singleThread`
* `--pool forks` as suggested in https://vitest.dev/guide/debugging.html#intellij-idea

I'd prefer to not pass any CLI options (i.e. take the 1st way), unless you know some scenarios where it might not work.

Also, it'd be great if you could postpone releasing Vitest 1.0.0 until IntelliJ / WebStorm 2023.3.1 with the fix is released (ETA: late December). Is it possible? Thanks.

**Comment by sheremet-va** [comment]
> Also, it'd be great if you could postpone releasing Vitest 1.0.0 until IntelliJ / WebStorm 2023.3.1 with the fix is released (ETA: late December). Is it possible? Thanks.

No, it is not possible. We made the change to `--pool` two months ago, this is plenty of time.

> What would the correct way to start Vitests for debugging?

`--poolOptions.threads.singleThread` uses a single `Worker` to run tests, `--pool forks` is _almost_ the same as `--no-threads`, but now it uses several forked processes to run tests. So if the problem is parallelism, this flag will not help. Maybe we can add `--debug` option to force Vitest to run a single process/thread/whatever it is?

**Reply by segrey** [reply]
I wish I knew about this change two months ago. What would be the best way to get notified about such changes in future?

It seems like there is no problem with parallelism and it works even with no options passed - at least it works for me on a few simple projects. However, as I can see, [guide/debugging](https://vitest.dev/guide/debugging.html#intellij-idea) suggests to pass `--pool forks`. Is there a reason for such advice? Maybe there are more complex projects or specific configuration where debugging don't works as expected unless parallelism is disabled. I'd be grateful if you could share more info on this matter.

**Reply by AriPerkkio** [reply]
You'll definitely want to use single worker/process when debugging. Otherwise your debugger will inconsistently jump from one test execution to another as they are run parallel. It's really up to the user if they want to use `pool: 'forks'` or `pool: 'threads'`. Some may even use `pool: 'vmThreads`. 

> What would be the best way to get notified about such changes in future?

As Vitest's previous versions were only `0.x` and beta releases, there were breaking changes happening a lot - intentionally. Now that 1.0 stable is being released it might be easier to follow breaking changes. 

I would also recommend IntelliJ to join [Vitest Ecosystem CI](https://github.com/vitest-dev/vitest-ecosystem-ci) by adding your own reference project that guarantees Vitest's `main` branch's changes do not break your use case.

**Reply by sheremet-va** [reply]
There is also an ecosystem channel on Discord. Are you there? We should probably notify everyone there when we are making a breaking change.

**Reply by segrey** [reply]
Regarding disabling parallelism when debugging, seems passing both `--poolOptions.threads.singleThread` and `--poolOptions.forks.singleFork` is the only option now. However, there are downsides according to warnings in their docs ([singleThread](https://vitest.dev/config/#pooloptions-threads-singlethread) / [singleFork](https://vitest.dev/config/#pooloptions-forks-singlefork)).

How about introducing a new option, like @sheremet-va suggested or maybe disable parallelism automatically based on [inspector.url()](https://nodejs.org/api/inspector.html#inspectorurl) != null? What do you think?

**Reply by AriPerkkio** [reply]
How exactly is IntelliJ passing the inspector arguments? Does it use Vitest's [`--inspect / --inspect-brk` arguments](https://vitest.dev/guide/debugging.html#node-inspector-e-g-chrome-devtools)? 

Maybe we could force single process/worker internally when it's possible to detect user is debugging. I guess this would be breaking change though. Currently Vitest simply throws an error when `--inspect` is passed without specifying single process/worker explicitly.

For now IntelliJ should pass all `--poolOptions.threads.singleThread --poolOptions.vmThreads.singleThread --poolOptions.forks.singleFork` flags so that it works on all pools.

**Reply by sheremet-va** [reply]
> For now IntelliJ should pass all `--poolOptions.threads.singleThread --poolOptions.vmThreads.singleThread --poolOptions.forks.singleFork` flags so that it works on all pools.

There is also `vmForks` planned - I wonder if we can unify similar options without making it confusing like before.

**Reply by segrey** [reply]
>How exactly is IntelliJ passing the inspector arguments?

By default, IntelliJ injects an agent inside a running Node.js process using `NODE_OPTIONS=--require /path/to/intellij/debugConnector.js`. The agent calls [inspector.open](https://nodejs.org/api/inspector.html#inspectoropenport-host-wait) and reports host/port to IntelliJ, so IntelliJ is able to connect and start a debug session. This approach allows to debug child processes as long as `NODE_OPTIONS` is inherited.

>Does it use Vitest's --inspect / --inspect-brk arguments?

No. As I can see, these options do a [similar thing](https://github.com/vitest-dev/vitest/blob/114a993c002628385210034a6ed625195fcc04f3/packages/vitest/src/runtime/inspector.ts#L16-L25), but it's a bit unclear how IntelliJ can connect to the opened port automatically.

> For now IntelliJ should pass all `--poolOptions.threads.singleThread --poolOptions.vmThreads.singleThread --poolOptions.forks.singleFork` flags so that it works on all pools.

Yep, that's the plan, thanks for confirming. The only problem is that seems `--poolOptions.vmThreads.singleThread` is widely supported (it's not mentioned in [the docs](https://vitest.dev/config/#pooloptions-vmthreads)). Probably, it should be checked when throwing [the abovementioned error](https://github.com/vitest-dev/vitest/blob/cb87f44e7682e4933c201d2552b647f5b2ae43a4/packages/vitest/src/node/config.ts#L116-L124), but I'm not sure. Anyway, IntelliJ will pass it too for completeness, thanks.

> Maybe we could force single process/worker internally when it's possible to detect user is debugging. I guess this would be breaking change though.

Yes, a breaking change. On the other hand, curious why people may want run tests with parallelism with open inspector. Seems like a misconfiguration to me.

I guess we have two things to discuss:
(a) Have a way to debug tests in isolated environment (because passing `--poolOptions.threads.singleThread`, `--poolOptions.forks.singleFork` and probably `--poolOptions.vmThreads.singleThread` runs tests in global context).
(b) Detect automatically that the main Vitest process is being debugged and disable parallelism automatically - run tests sequentially and in isolated environment.
I'd be happy to file corresponding issues if you confirm.

**Reply by sheremet-va** [reply]
> Have a way to debug tests in isolated environment (because passing `--poolOptions.threads.singleThread`, `--poolOptions.forks.singleFork` and probably `--poolOptions.vmThreads.singleThread` runs tests in global context).

I don't think running `singleThread`/`singleFork` without the user's consent is a good idea since it alters the environment dramatically, and some code is not meant to run that way. Passing down `--poolOptions.threads.maxThreads=1` (and others) should have the same effect but it will run tests isolated (which will make it much slower, but at the same speed `--no-threads` used to run in 0.36.4). 

The only problem for me is the need to pass it down for each possible pool - this looks like a poor API for tools that don't have the knowledge of the current pool (like JetBrains extension)

**Reply by segrey** [reply]
Great, thanks for the suggestion. Passing `--poolOptions.threads.maxThreads=1` only leads to `RangeError: options.minThreads and options.maxThreads must not conflict`. Same error for `pool: 'forks'` and `pool: 'vmThreads'`.

It works fine if both `--poolOptions.threads.maxThreads=1` and `--poolOptions.threads.minThreads=0` are passed. Is it OK?

**Reply by sheremet-va** [reply]
Yeah, you need to pass down both. I think setting them both to 1 should be allright.

**Reply by sheremet-va** [reply]
> The only problem for me is the need to pass it down for each possible pool - this looks like a poor API for tools that don't have the knowledge of the current pool (like JetBrains extension)

Maybe we can have `--maxWorkers`, `--minWorkers` and `--isolatedWorkers` that will override every compatible option in every pool?

**Reply by segrey** [reply]
Indeed, `--poolOptions.threads.minThreads=1` works fine, thanks.

>Maybe we can have `--maxWorkers`, `--minWorkers` and `--isolatedWorkers` that will override every compatible option in every pool?

Looks good to me. Is `--isolatedWorkers` a boolean flag or a number?

**Reply by sheremet-va** [reply]
A boolean. The same as `--poolOptions.threads.isolate`. If it was a number, what would you expect there? Maybe this is a poor choice for the name

**Reply by segrey** [reply]
If it would be a number, I would assume that it's amount of isolated workers, other workers are not isolated (I know, insane 😄).
Maybe we don't need a boolean flag enabling isolation? Just a flag disabling it: `--disableIsolation`?

**Reply by segrey** [reply]
Let's summarize how IntelliJ 2023.3.1 should launch Vitest for debugging. If I got it right, IntelliJ should pass the following six flags (regardless of the actual pool, because it is not known to the IDE):
```
--poolOptions.threads.maxThreads=1
--poolOptions.threads.minThreads=1

--poolOptions.forks.maxForks=1
--poolOptions.forks.minForks=1

--poolOptions.vmThreads.maxThreads=1
--poolOptions.vmThreads.minThreads=1
```
I've tested this approach and it works fine for me.

What's next? When `--maxWorkers` and `--minWorkers` are available, IntelliJ will migrate to it.

>(b) Detect automatically that the main Vitest process is being debugged and disable parallelism automatically - run tests sequentially and in isolated environment.

Implementing this approach would not only result in zero configuration for tools (like IDEs), but also would allow debugging right from the terminal (where extra Vitest option cannot be passed, but it's possible to open inspector using the agent injected via `NODE_OPTIONS`).

**Reply by sheremet-va** [reply]
> If I got it right, IntelliJ should pass the following six flags (regardless of the actual pool, because it is not known to the IDE):

Currently, yes.

> What's next? When --maxWorkers and --minWorkers are available, IntelliJ will migrate to it.

If we decide to add it. @AriPerkkio what do you think?

> Implementing this approach would not only result in zero configuration for tools (like IDEs), but also would allow debugging right from the terminal (where extra Vitest option cannot be passed, but it's possible to open inspector using the agent injected via `NODE_OPTIONS`).

I am personally not sure if there is a reliable way to do this (is inspector always open only for debugging?), so I would trust @AriPerkkio here 😄

**Reply by AriPerkkio** [reply]
> > When `--maxWorkers` and `--minWorkers` are available, IntelliJ will migrate to it.
>
> If we decide to add it.

I think using these flags as shorthands for other `--poolOptions.<name>.<option>` would be ok.

> I am personally not sure if there is a reliable way to do this (is inspector always open only for debugging?)

We could disable parallelism when we detect that **Vitest's** `--inspect` flags are used. But I'm not yet sure for other resolving logics, e.g. reading `NODE_OPTIONS`.

**Reply by AriPerkkio** [reply]
> (b) Detect automatically that the main Vitest process is being debugged and disable parallelism automatically

Just remembered that I actually do use inspector even when multi threading or multi processing is used, when doing performance and memory profiling. Inspector usage is not always about debugger breakpoints.

**Reply by segrey** [reply]
Agree, profiler also enables inspector. Okay, it won't work easily.
Maybe instead of adding `--maxWorkers` and `--minWorkers` it would be more natural to have a boolean flag `--disableParallelism` or something else  like `--runInBand` in Jest?

**Reply by sheremet-va** [reply]
We don't really add `disable` flags, but we can add `--no-parallelism` that just sets maxWorkers and minWorkers to 1?

**Reply by AriPerkkio** [reply]
I like `--no-parallelism`. Let's add that one.

**Reply by segrey** [reply]
Perfect! Looking forward to `--no-parallelism`.

**Comment by reuel88** [comment]
I've read these comments over and over again and still don't understand what's going on. I just want to be able to click the debug icon on the side and have it debug like before. Is there a way to do that?

**Comment by MikeDabrowski-EQM** [comment]
2025 versions of IntelliJ and still broken. I am able to get the debugger connected adding `--inspect-brk` but it has caveats on its own like the process not being killed and blocking the debug port or additional 'breakpoints' from various 'workers' stopping the execution. By no means a workaround :(

**Comment by justspamjustin** [comment]
It looks like the issue was reported to intellij and is now fixed: https://youtrack.jetbrains.com/issue/WEB-75404/Vitest-v4-debug-fails-with-Unknown-option-poolOptions
